// detox-copilot/dist/actions/StepPerformer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StepPerformer = void 0;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const crypto = __importStar(require("crypto"));
class StepPerformer {
    context;
    promptCreator;
    codeEvaluator;
    snapshotManager;
    promptHandler;
    cache = new Map();
    cacheFilePath;
    constructor(context, promptCreator, codeEvaluator, snapshotManager, promptHandler, cacheFileName = 'detox_copilot_cache.json') {
        this.context = context;
        this.promptCreator = promptCreator;
        this.codeEvaluator = codeEvaluator;
        this.snapshotManager = snapshotManager;
        this.promptHandler = promptHandler;
        this.cacheFilePath = path.resolve(process.cwd(), cacheFileName);
    }
    generateCacheKey(step, previous, viewHierarchy) {
        const viewHierarchyHash = crypto.createHash('md5').update(viewHierarchy).digest('hex');
        return JSON.stringify({ step, previous, viewHierarchyHash });
    }
    loadCacheFromFile() {
        try {
            if (fs.existsSync(this.cacheFilePath)) {
                const data = fs.readFileSync(this.cacheFilePath, 'utf-8');
                const json = JSON.parse(data);
                this.cache = new Map(Object.entries(json));
            }
            else {
                this.cache.clear(); // Ensure cache is empty if file doesn't exist
            }
        }
        catch (error) {
            console.warn('Error loading cache from file:', error);
            this.cache.clear(); // Clear cache on error to avoid stale data
        }
    }
    saveCacheToFile() {
        try {
            const json = Object.fromEntries(this.cache);
            fs.writeFileSync(this.cacheFilePath, JSON.stringify(json, null, 2), { flag: 'w+' });
        }
        catch (error) {
            console.error('Error saving cache to file:', error);
        }
    }
    async perform(step, previous = []) {
        // todo: replace with the user's logger
        console.log("\x1b[90m%s\x1b[0m%s", "Copilot performing: ", `"${step}"`);
        // Load cache before every operation
        this.loadCacheFromFile();
        const snapshot = this.promptHandler.isSnapshotImageSupported() ? await this.snapshotManager.captureSnapshotImage() : undefined;
        const viewHierarchy = await this.snapshotManager.captureViewHierarchyString();
        const isSnapshotImageAttached = snapshot != null && this.promptHandler.isSnapshotImageSupported();
        const cacheKey = this.generateCacheKey(step, previous, viewHierarchy);
        if (this.cache.has(cacheKey)) {
            const cachedPromptResult = this.cache.get(cacheKey);
            return this.codeEvaluator.evaluate(cachedPromptResult, this.context);
        }
        const prompt = this.promptCreator.createPrompt(step, viewHierarchy, isSnapshotImageAttached, previous);
        let promptResult;
        try {
            promptResult = await this.promptHandler.runPrompt(prompt, snapshot);
            const result = await this.codeEvaluator.evaluate(promptResult, this.context);
            // Cache the result
            this.cache.set(cacheKey, promptResult);
            this.saveCacheToFile();
            return result;
        }
        catch (error) {
            // Extend 'previous' array with the failure message
            const failedAttemptMessage = promptResult
                ? `Failed to evaluate "${step}", tried with generated code: "${promptResult}". Should we try a different approach? If can't, return a code that throws a descriptive error.`
                : `Failed to perform "${step}", could not generate prompt result. Should we try a different approach? If can't, return a code that throws a descriptive error.`;
            const newPrevious = [...previous, {
                    step,
                    code: failedAttemptMessage,
                    result: undefined,
                }];
            const retryCacheKey = this.generateCacheKey(step, newPrevious, viewHierarchy);
            if (this.cache.has(retryCacheKey)) {
                const cachedRetryPromptResult = this.cache.get(retryCacheKey);
                return this.codeEvaluator.evaluate(cachedRetryPromptResult, this.context);
            }
            const retryPrompt = this.promptCreator.createPrompt(step, viewHierarchy, isSnapshotImageAttached, newPrevious);
            try {
                const retryPromptResult = await this.promptHandler.runPrompt(retryPrompt, snapshot);
                const retryResult = await this.codeEvaluator.evaluate(retryPromptResult, this.context);
                // Cache the result under the original cache key
                this.cache.set(cacheKey, retryPromptResult);
                this.saveCacheToFile();
                return retryResult;
            }
            catch (retryError) {
                // Throw the original error if retry fails
                throw error;
            }
        }
    }
}
exports.StepPerformer = StepPerformer;
